Code/GraphMol/FileParsers/PNGParser.cpp

```diff
@@ -0,0 +1,293 @@
+//
+//  Copyright (C) 2020 Greg Landrum
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+// details of how to handle the PNG file taken from OpenBabel's PNG handling
+// code:
+// https://github.com/openbabel/openbabel/blob/master/src/formats/pngformat.cpp
+
+#include "PNGParser.h"
+#include <GraphMol/MolPickler.h>
+#include <RDGeneral/FileParseException.h>
+#include <RDGeneral/StreamOps.h>
+#include <vector>
+#include <boost/crc.hpp>
+#include "FileParsers.h"
+#ifdef RDK_USE_BOOST_IOSTREAMS
+#include <zlib.h>
+#endif
+
+namespace RDKit {
+
+namespace PNGData {
+const std::string smilesTag = "rdkitSMILES";
+const std::string molTag = "rdkitMOL";
+const std::string pklTag = "rdkitPKL";
+}  // namespace PNGData
+
+namespace {
+std::vector<unsigned char> pngHeader = {137, 80, 78, 71, 13, 10, 26, 10};
+bool checkPNGHeader(std::istream &inStream) {
+  for (auto byte : pngHeader) {
+    unsigned char ibyte;
+    inStream.read((char *)&ibyte, 1);
+    if (ibyte != byte) {
+      return false;
+    }
+  }
+  return true;
+}
+}  // namespace
+
+std::map<std::string, std::string> PNGStreamToMetadata(std::istream &inStream) {
+  // confirm that it's a PNG file:
+  if (!checkPNGHeader(inStream)) {
+    throw FileParseException("PNG header not recognized");
+  }
+  std::map<std::string, std::string> res;
+  // the file is organized in chunks. Read through them until we find the tEXt
+  // block FIX: at some point we'll want to also include zEXt here, but that
+  // requires zlib
+  while (inStream) {
+    std::uint32_t blockLen;
+    inStream.read((char *)&blockLen, sizeof(blockLen));
+    if (inStream.fail()) {
+      throw FileParseException("error when reading from PNG");
+    }
+    // PNG is big endian, make sure we handle the order correctly
+    blockLen = EndianSwapBytes<BIG_ENDIAN_ORDER, HOST_ENDIAN_ORDER>(blockLen);
+    char bytes[4];
+    inStream.read(bytes, 4);
+    if (inStream.fail()) {
+      throw FileParseException("error when reading from PNG");
+    }
+    auto beginBlock = inStream.tellg();
+    if (bytes[0] == 'I' && bytes[1] == 'E' && bytes[2] == 'N' &&
+        bytes[3] == 'D') {
+      break;
+    }
+#ifndef RDK_USE_BOOST_IOSTREAMS
+    bool alreadyWarned = false;
```

Comment:
I'm assuming you want this to be static.

Reply:
hmm, I don't think it's necessarily bad to warn once per molecule. This shouldn't happen all that often: the conda builds, which account for the majority of users, are all built against boost::iostreams.